baseapp/handlers.go

```diff
@@ -0,0 +1,34 @@
+// Copyright 2020 Palantir Technologies, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package baseapp
+
+import (
+	"net/http"
+
+	"go.opentelemetry.io/otel/api/trace"
+	"go.opentelemetry.io/otel/instrumentation/othttp"
+)
+
+// TelemetryHandler wraps `othttp.WithRouteTag` to use the route tag value as the name of the span. By default the span
+// name will be the name of the service defined in the exporters.
+func TelemetryHandler(route string, h http.Handler) http.Handler {
```

Comment:
This feels like middleware to me, since it takes a handler and returns a handler. Looking at your example usage, I understand why it's a handler (the route changes on each call), but given `go-baseapp` has some opinions, it seems like we could make this friendlier by auto-extracting the route pattern from Goji.

In particular, you can use [`middleware.Pattern`](https://godoc.org/goji.io/middleware#Pattern) to get the matched pattern and then call `String()` if it exists to get the route (`pat.Pattern` implements `String` but not all implementations will). If the pattern doesn't exist, maybe the route can be inferred from the request URI.

One downside is that if I remember correctly, patterns from submuxes are not added together. The matched pattern will not include whatever prefix was used to register the submux. I have some internal code to deal with this, but there might other ways to handle it too.

Reply:
Sadly Goji is severely lacking here. The match pattern only returns the root mux and none of the sub muxes, as you mentioned. Furthermore, it doesn't actually give you back the actual pattern, just the result of the match. So, for example, a pattern of `/role/:name`, when requested with `/role/some-role` with return `/role/some-role` which goes against the [spec for OpenTelemetry](https://github.com/open-telemetry/opentelemetry-specification/blob/master/specification/trace/semantic_conventions/http.md#name).

By default if you don't use this handler it will fall back to the the service name, which is more preferable over a span name with the URL in it

Reply:
An alternative, which I'll push up in one complete commit in case we want to revert, is to extend the Goji mux with a new `HandleWithTelemetry` function. While not as configurable as this handler, it would allow for a relatively simple switch without needing to know what happens underneath